test/test_testing.py

```diff
@@ -2396,6 +2396,145 @@ def test_begin_with_initial_hooks_install_sets_status(self):
+        harness.charm.changes = []  # Clean up records to have a fresh count for `upgrade`
+        harness.upgrade()
+        self.assertEqual(
+            harness.charm.changes,
+            [
+                # {'name': 'stop'},  # the "stop" record is dropped due to charm reinit on upgrade
+                {'name': 'upgrade-charm'},
+                {'name': 'config-changed', 'data': {}},
+                {'name': 'start'},
+            ]
+        )
+
+    def test_upgrade_with_one_relation(self):
+        """No relation hooks should fire on upgrade."""
+        class CharmWithDB(RelationEventCharm):
+            def __init__(self, framework):
+                super().__init__(framework)
+                self.observe_relation_events('db')
+        harness = Harness(CharmWithDB, meta='''
+            name: test-app
+            requires:
+              db:
+                interface: sql
+            ''')
+        self.addCleanup(harness.cleanup)
+        harness.set_leader()
+        rel_id = harness.add_relation('db', 'postgresql')
+        harness.add_relation_unit(rel_id, 'postgresql/0')
+        harness.update_relation_data(rel_id, 'postgresql/0', {'new': 'data'})
+        harness.begin_with_initial_hooks()
+        harness.charm.changes = []  # Clean up records to have a fresh count for `upgrade`
+        harness.upgrade()
+        self.assertEqual(
+            harness.charm.changes,
+            [
+                # {'name': 'stop'},  # the "stop" record is dropped due to charm reinit on upgrade
+                {'name': 'upgrade-charm'},
+                {'name': 'config-changed', 'data': {}},
+                {'name': 'start'},
+            ]
+        )
+
+    def test_upgrade_with_multiple_units(self):
+        """No relation hooks should fire on upgrade."""
+        class CharmWithDB(RelationEventCharm):
+            def __init__(self, framework):
+                super().__init__(framework)
+                self.observe_relation_events('db')
+        harness = Harness(CharmWithDB, meta='''
+            name: test-app
+            requires:
+              db:
+                interface: sql
+            ''')
+        self.addCleanup(harness.cleanup)
+        harness.set_leader()
+        rel_id = harness.add_relation('db', 'postgresql')
+        harness.add_relation_unit(rel_id, 'postgresql/0')
+        harness.add_relation_unit(rel_id, 'postgresql/1')
+        harness.begin_with_initial_hooks()
+        harness.charm.changes = []  # Clean up records to have a fresh count for `upgrade`
+        harness.upgrade()
+        self.assertEqual(
+            harness.charm.changes,
+            [
+                # {'name': 'stop'},  # the "stop" record is dropped due to charm reinit on upgrade
+                {'name': 'upgrade-charm'},
+                {'name': 'config-changed', 'data': {}},
+                {'name': 'start'},
+            ]
+        )
+
+    def test_upgrade_status_consistency(self):
```

Comment:
Not sure why we're testing this. In general it's not true, is it? An upgrade could in principle introduce non-backwards compatible changes (e.g. add a mandatory config option...)